pkg/kubelet/status/status_manager.go

```diff
@@ -463,6 +464,15 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 		}
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		// Make sure all deleted, Pending pods are transitioned to a terminal phase.
+		if pod.DeletionTimestamp != nil && status.Phase == v1.PodPending {
```

Comment:
Yes, this is the plan, we will be able to adjust the condition in the other PR which deals with Running pods: https://github.com/kubernetes/kubernetes/pull/116324

Reply:
So, one consideration raised by @bobbypage was that if all containers are completed with 0 exit code, then the pod status should be Succeeded. Thus, we handle the running pods independently, in a separate PR.

Reply:
Hrm.  I don't understand why this is two PRs.  This method TerminatePod is called after all containers have exited (normally or abnormally) - by the time we get to this method any information about the pod from the runtime status has already been calculated.  It is impossible after this point for any additional information to change our decision about the pod's phase (that's what generateAPIPodStatus in syncTerminatedPod does).

So from a behavior perspective, there is no reason we could be in Running (other than  a bug in generateAPIPodStatus), there should be no reason to be in Unknown, or Pending.  It has nothing to do with deletion even - to get to this point, you should either have already transitioned to a terminal phase (based on normal operation) or something went wrong.  This is correct point in the kubelet to enforce that transition.

I would probably treat the scenarios with a switch:

```
switch status.Phase {
case PodSucceeded, PodFailed:
   // do nothing, already terminal
case PodPending, PodUnknown:
   ... // your case
case PodRunning:
  klog.InfoS("Pod should have already been marked as terminal, programmer error")
  status.Phase = PodFailed
default:
  klog.ErrorS("Unknown phase, programmer error")
  status.Phase = PodFailed
}
```

Reply:
Note the reason why this didn't already do it is prior to the improvements to the pod worker, we didn't know for certain the containers were stopped at this point or why.  Now we do, and this is kind of the last bit of that transition to enforce.

Reply:
I see, I'm happy to combine the PRs, will progress on it now, and close the other.

Reply:
> So, one consideration raised by @bobbypage was that if all containers are completed with 0 exit code, then the pod status should be Succeeded. 

That calculation should be happening already at the beginning of syncTerminatedPod - generate phase should have all container status information already because we should be waiting to get the latest container status from the runtime in pod_workers (the `status, err = p.podCache.GetNewerThan(pod.UID, lastSyncTime)` check), then we invoke syncTerminatedPod with that latest value.  Then TerminatePod gets invoked with the latest status, so any succeeded pods should already be succeeded.

Reply:
This should be fine, as the `syncTerminatingPod` calls `generateAPIPodStatus`, which calls `getPhase` where I put the computation now, after merge.

Reply:
Interesting, there is one test for mirror pods which fails when `TerminatePod` transitions all pods (as in the snippet), without checking for `deletionTimestamp!=nil`. I'm not yet sure if we can adjust the assert or this is relevant.

Reply:
For now, I scope the change to transition only deleted pods, alternatively I could scope it to non mirror pods, using `kubetypes.IsMirrorPod`, not should what is cleaner.